Type stats for units and buildings and refactoring

MaanooAk opened the pull request:
- Data structures
  - struct **ObjectTypeStats**
    -  enum **AttackType**
    - enum **GarrisonType**
    - struct **PopulationCount**
  - struct **UnitTypeStats** :  ObjectTypeStats
  - struct **BuildingTypeStats** : ObjectTypeStats
  - struct **GameData**
- Data
  - **LoadUnitTypeStats()**
    - helper inline functions
  - **LoadBuildingTypeStats()**
    - helper inline functions
- Tests
  - DataValidation

MaanooAk commented on src/FreeAge/common/type_stats_data.cpp:
```diff
@@ -0,0 +1,330 @@
+
+// NOTE: The following functions should be able to inline all the helper function calls and elimenate
+//       any redundant value sets.
+
```
@puzzlepaint I added some of the units and buildings with this design of using helper inline functions. I use comments to make more clear what every value means, obviously this kind of format would be only for this file, which I think its ok because its not really code but just data. 
I know two people is basically impossible to agree in one format, but before I add all the current supported units: what do you think about how it looks?

The idea is to use this functions instead of just setting every value in order to avoid making mistakes and also compressing the size of the code.

puzzlepaint replied:
+1 for the inline comments, they seem helpful here (in case the IDE cannot show the parameter names inline).

For me it would seem more natural to make the `SetXY()` functions member functions of the Stats structs. The current style seems more like C than C++ ;) But it does not really matter much.

MaanooAk replied:
I think to avoid having all the getters/setters we could just pass everywhere `const Stats` and only the load code and `Modification` code have access the no const (I think only Modification should be able to change the stats after the initial set up)

I know there is the _friend_ declaration to make setters work only in some places but I'm not sure how it works and if it's worth it...

puzzlepaint replied:
I am not sure whether I understand the concern here. With something like:
```c++
struct Stats {
  int attribute1;
  int attribute2;

  void SetAllUnitAttributes(int attribute1, int attribute2);
};
```
passing a `const Stats&` around would allow to read the attributes without the need for getters, and passing a non-const `Stats` around would allow to set the attributes without the need for a friend class/struct?

MaanooAk replied:
So the `SetUnitBasic` would be "visible" everywhere but you could not use it because you have a "const" reference... I understand now. I have been working mostly with Java the last months, where something like this is not possible.

However all the Set functions will only ever be needed in the `type_stats_data.cpp` (and even be removed entirely if the data in the future are loaded from files).

By the way I thought `SetXY` meant having a setter for each field... this probably why you said XY instead of X.

puzzlepaint replied:
Sorry for not being clear enough. Yes, this would make the functions visible outside of `type_stats_data.cpp`. I am totally fine with both implementations, the more object-oriented one just "feels" a bit more natural because it keeps the code interacting with the structs closer to the structs themselves.

MaanooAk replied:
After a bit of web searching: the setters could be private and the `LoadUnitTypeStats` could be declared as a friend, but then all the Set* would either have to be put on the hpp (which would pobably be included by a lot of files) or remove the `inline` keyword and put them to a cpp. I will leave it like this for now and in any case a future refactor would only be a couple of easy find-and-replace operations.

puzzlepaint commented on src/FreeAge/common/type_stats.hpp:
```diff
@@ -0,0 +1,140 @@
+struct UnitTypeStats : public ObjectTypeStats {
+
+  /// The population space needed.
+  PopulationDemand populationDemand;
```
Maybe this can be generalized in `ObjectTypeStats` such as to unify `UnitTypeStats::populationDemand` and `BuildingTypeStats::populationSpace`. The unified attribute could for example be positive if the object gives population space, and negative if the object takes up population space.

Buildings taking up population space need to be handled anyway due to the Feitoria building of the Portuguese, which takes up 20 population.

MaanooAk replied:
This a good idea. How could we store the half population then? I really don't like having *2 everywhere, but it seems like the only solution... Any ideas?

puzzlepaint replied:
One idea would be making a `PopulationCount` class which has a descriptive interface, handles the *2 or /2 internally, and otherwise behaves like a number. It could have functions like `SetToOneHalf()`, `SetToInteger(int value)`, ...

MaanooAk replied:
I knew you would have a the perfect solution! I will add this!

grst commented on src/FreeAge/common/type_stats.hpp:
```diff
@@ -0,0 +1,140 @@
+enum class GarrisonType {
+  None,
+  Units, // TransportShip ?
+  VillagersAndFootArchersAndInfantry, // TownCenter and towers
```
More compact `VillagersAndFootSoldiers`? It applies also for Rams I think. 
There's also the Siege Tower, which, according to [aoewiki](https://ageofempires.fandom.com/wiki/Garrison) can only garrison FootSoldiers but no vills.

grst replied:
There's more nasty details, e.g. monks and petards cannot garrison in siege rams/towers.

MaanooAk replied:
I wiil add `VillagersAndFootSoldiers` and `FootSoldiers`, but I don't think the monks and petards will cause any problem because they are not considered soldiers

grst replied:
But at least monks can garrison in TCs and towers can't they? 
But probably these details are also irrelevant at this stage of the project, so feel free to ignore.

MaanooAk replied:
Yes, you are right! No, I don't think they are irrelevant, they help having to revisit and refactor code in later stages (which could be painful sometimes).

Then I think they could be considered "Villagers" in the context of garrison (I'm not sure for that either, does it work?)

grst replied:
> Then I think they could be considered "Villagers" in the context of garrison (I'm not sure for that either, does it work?)

Not really, as vills can garrison in rams, while monks can't. 

Just found this page, it seems like a great resource, not only for garrisoning: 
https://www.aoe2sheet.info/knowledge_base.html#garrisoning

OT: 
Holy shit, have a look at the monk mechanics!

MaanooAk replied:
@grst How about now?
```c++
enum class GarrisonType {
  /// No unit can be garrison.
  None,
  /// All units
  AllUnits, // Transport ships
  /// Villagers, monks and foot soldiers only.
  VillagersAndMonksAndFootSoldiers, // TownCenter and towers
  /// Villagers and foot soldiers only.
  VillagersFootSoldiers, // Rams and Siege Towers
  /// All units except siege units.
  NonSiege, // Castle
  /// Villagers only.
  Villagers, // Khmer houses
  /// Units from production buildings can be trained directly into garrison,
  /// but not garrison from outside.
  Production,
  /// Same as Production plus Relics.
  ProductionAndRelics
};
```

> Holy shit, have a look at the monk conversion-mechanics!

Don't worry it looks complicated but its mainly only 3 variables, on the wiki is explained more from the code perspective.

grst replied:
Looks great now!

grst commented on src/FreeAge/common/type_stats_data.cpp:
```diff
@@ -0,0 +1,330 @@
+    UnitTypeStats& s = InitUnit(unitTypeStats, UnitType::FemaleVillager);
+
+    SetUnitBasic(s,
+      /* hp     */ 25,
```
Personally, I don't like that the argument is only defined by the order. Looks like a great source of copy&paste errors when completing the unit stats. I have literally no experience with C++ so I don't know what options there are -- no named arguments as far as I can tell. 

Would a chained, object-oriented approach work? Sth like: 

```c++
s = InitUnit(unitTypeStats, UnitType::FemaleVillager)
          .setHp(25)
          .setArmor(0,0)
          .setRadius(1)
```

puzzlepaint replied:
This would work, but in contrast to a function like `SetUnitBasic()`, it would not directly ensure that all of the relevant options are set together (so it would be possible to forget setting a new option for some unit, which may then remain uninitialized). It would be possible to add a mechanism that checks for a complete initialization, but that would also add some complexity.

MaanooAk replied:
You example could be implemented, but the reason of the functions is to ensure that some info is given grouped for example:
```c++
void SetUnitMeleeAttack(UnitTypeStats& s, 
  float fireRate, int meleeDamage, float attackDelay)
```
sets everything that is needed for a melee attack, and sets more things than the 3 given params. It also enusres eg. that if you have a melee atack you have explicity set the fireRate and not left the default value.

I understand that there is a risk that you reorder things believing its not a functions, but I think is risk worth taking.

MaanooAk replied:
( I didn't want to repeat @puzzlepaint, his answer had not loaded when I wrote mine )

puzzlepaint replied:
Thinking about this, I came up with an idea for "simulating" named parameters in C++:

```c++
template <typename T>
class NamedParameterBase {
 public:
  inline NamedParameterBase(T value)
      : value(value) {}
  
  inline operator T() { return value; }

 private:
  T value;
};

#define MAKE_NAMED_PARAMETER(ParamName, ParamType) \
    class ParamName : public NamedParameterBase<ParamType> { using NamedParameterBase<ParamType>::NamedParameterBase; }

MAKE_NAMED_PARAMETER(FireRate, float);
MAKE_NAMED_PARAMETER(MeleeDamage, int);
MAKE_NAMED_PARAMETER(AttackDelay, float);

#undef MAKE_NAMED_PARAMETER

// Then one can define:
void SetUnitMeleeAttack(UnitTypeStats& s, 
  FireRate fireRate, MeleeDamage meleeDamage, AttackDelay attackDelay);

// ... which is later used like this:
SetUnitMeleeAttack(
    FireRate(0.4f),
    MeleeDamage(2),
    AttackDelay(0.3));
```

This does add some (but not much) code overhead. It might be an idea for formalizing the inline comments and prevents mixing up the order as suggested by @grst.

MaanooAk replied:
You still cannot reorder the params but it will give an error, right? How you then get the value out, with `fireRate()` ?

Could this use [User-defined literals](https://en.cppreference.com/w/cpp/language/user_literal) ?

puzzlepaint replied:
Yes, this is just for having (and thus verifying) the parameter names in the function calls. it does not make reordering possible.

In the example code above, getting the value out would work with the conversion operator that is defined in NamedParameterBase. So, for a float parameter for example, in a context where the parameter is directly converted to a float, it could just be used like one. For example, this works: `float test = attackDelay;` This will not work though if the conversion is not clear, then something like `static_cast<float>(attackDelay)` would need to be used, I think. An alternative would be to make a simple `get()` function.

I haven't used user-defined literals before, but I think this should work with them as well, yes. It's probably just a question of which syntax seems preferable. Either:
```c++
SetUnitMeleeAttack(
    FireRate(0.4f),
    MeleeDamage(2),
    AttackDelay(0.3));
```
or:
```c++
SetUnitMeleeAttack(
    0.4_fireRate,
    2_meleeDamage,
    0.3_attackDelay);
```

MaanooAk replied:
This is very clever, but I dont know... :)

```c++
    SetUnitBasic(s,
      /* max hp */ 40,
      /* armor  */ 0,  1,
      /* radius */ 1,
      /* speed  */ 0.9,
      /* los    */ 4);
```
```c++
   SetUnitBasic(s,
      MaxHp(40),
      MArmor(0), 
      PArmor(1),
      Radius(1),
      Speed(0.9),
      Los(4));
```
```c++
    SetUnitBasic(s,
      MaxHp  (40),
      MArmor (0), 
      PArmor (1),
      Radius (1),
      Speed  (0.9),
      Los    (4));
```

> I know two people is basically impossible to agree in one format

I think we have reached this state

puzzlepaint replied:
> > I know two people is basically impossible to agree in one format
> 
> I think we have reached this state

Then, given that you currently put in the most effort by far, I think that you should decide :)

MaanooAk commented on src/FreeAge/common/type_stats.hpp:
```diff
@@ -154,3 +154,11 @@ struct BuildingTypeStats : public ObjectTypeStats {
 };
+
+// TODO: move elsewhere
+struct GameData {
```
This is **temp** struct to keep all the data (currently the UnitTypeStats and the BuildingTypeStats).

MaanooAk commented on src/FreeAge/common/player.hpp:
```diff
@@ -135,3 +136,36 @@ struct PlayerStats {
   bool buildingExisted[static_cast<int>(BuildingType::NumBuildings)];
 };
+
+struct Player {
```
@puzzlepaint Take look at this. Basically have this base Player class for both the server and the client for keeping track all the type stats. Server will have all the real values and client will have the current players real values and the latest know for each of the other players. Then access to the stats would be 
```c++
players.at(playerIndex).GetUnitStats(UnitType::Scout).fireRate
```
Or this could be wrapped in the Client/SeverUnit/Building
```c++
unit->GetStats().fireRate
```
Also, you were right (again), there no need for a UnitTypeInfo. The AddModification could be added inside the Player and apply the effects directly to the data in the vectors (which is client/server independent).

puzzlepaint replied:
Sounds / looks good to me. A shortcut such as `unit->GetStats().fireRate` (perhaps not in the Unit class though as this might not have access to `players`) is probably highly desirable, since I guess that in many cases, the stats for some specific object are needed, and it would be quite cumbersome to always do `players.at(unit->GetPlayerIndex()).GetUnitStats(unit->GetType()).fireRate`.

MaanooAk replied:
Unit has a playerIndex, shouldn't be replaced with a `Player*`? It should be safe to use a pointer.

puzzlepaint replied:
Indeed, this seems like it should solve this nicely.

puzzlepaint commented on src/FreeAge/common/type_stats.hpp:
```diff
@@ -0,0 +1,164 @@
+    : doubledCount(0) {};
+  
+  inline void SetToIntegerPopulationSpace(int count) { doubledCount = count / 2; }
+  inline void SetToIntegerPopulationDamand(int count) { doubledCount = -count / 2; }
```
Damand -> Demand

puzzlepaint commented on src/FreeAge/common/player.hpp:
```diff
@@ -135,3 +136,36 @@ struct PlayerStats {
+  inline const UnitTypeStats& GetUnitStats(UnitType unitType) const { return unitTypeStats.at(static_cast<int>(unitType)); }
+  inline const BuildingTypeStats& GetBuildingStats(BuildingType buildingType) const { return buildingTypeStats.at(static_cast<int>(buildingType)); }
+
+  /// The player index.
```
Comment on what this index is for (which array(s) is this index used for, maybe mention `match->GetPlayerIndex`)?

MaanooAk commented:
So... I did things! Everything works. I will stop adding things now, its a good point.

I added the data for only the types that already existed, removed/replaced the old static get functions and I changed only the game logic that already existed.

MaanooAk commented on src/FreeAge/server/map.cpp:
```diff
@@ -47,8 +48,10 @@ ServerMap::~ServerMap() {
 
 
-void ServerMap::GenerateRandomMap(int playerCount, int seed) {
+void ServerMap::GenerateRandomMap(Game* game, int seed) {
```
I forgot about this... I basically pass the whole game (in the future could be just a struct with the players and the map settings). Players are need because the civ can affect the starting objects (number of villagers, starting scout unit, (possibly more things?))

MaanooAk commented on src/FreeAge/client/render_window.cpp:
```diff
@@ -2289,25 +2290,23 @@ void RenderWindow::RenderSelectionPanel(QOpenGLFunctions_3_2_Core* f) {
-      u32 maxHP;
-      if (singleSelectedObject->isUnit()) {
-        maxHP = GetUnitMaxHP(AsUnit(singleSelectedObject)->GetType());
+      int maxHP = singleSelectedObject->GetObjectStats().maxHp;
```
Probably more places in the code could use the `GetObjectStats()` to reduce code duplication, I didn't search for them.

MaanooAk commented on src/FreeAge/common/player.hpp:
```diff
@@ -50,18 +53,20 @@ struct PlayerStats {
-  /// TODO: implement on the client, now only used by the server
+  /// TODO: Implement on the client, now only used by the server. 
+  /// TODO: Encapsulate and use the PopulationCounter
   int populationInProduction = 0;
```
This should be properly implemented at some point...

MaanooAk commented on src/FreeAge/common/player.hpp:
```diff
@@ -50,18 +53,20 @@ struct PlayerStats {
 
-  inline int GetPopulationCount() const { return populationCount; }
-  inline int GetAvailablePopulationSpace() const { return availablePopulationSpace; }
+  inline int GetPopulationDemand() const { return doubledPopulationDemand / 2; }
```
I renamed from Count to Demand because the `PopulationCount` could be confused with `AliveUnitCount` (which could be impl in the future)

puzzlepaint commented on src/FreeAge/common/type_stats.hpp:
```diff
@@ -0,0 +1,167 @@
+  int doubledCount;
+};
+
+struct ObjectTypeStats {
```
Maybe it could make sense to add `// TODO: not used by the game yet` to each of the attributes that is not actually used by the game yet to keep an overview of that

MaanooAk commented on src/FreeAge/server/game.cpp:
```diff
@@ -1379,17 +1385,12 @@ bool Game::SimulateMeleeAttack(u32 /*unitId*/, ServerUnit* unit, u32 targetId, S
       target != nullptr) {
     // Compute the attack damage.
+    // TODO: handle all the attack types
+    CHECK_EQ(unit->GetStats().attackType == AttackType::Default, true);
```
Yes, CHECK_EQ does not work with enum classes, I was looking for CHECK_TRUE but I guess it's just CHECK.

It looks like CHECK_EQ cannot compile because the `operator<<` is missing from the AttackType. I have added a TODO suggestion comment (somewhere) about implementing a GetXInternaName where X all the enums (mainly for debugging and logging). With that the `operator<<` would be trivial to implement.

puzzlepaint replied:
Alternatively (with less pretty output though), one can just do `static_cast<int>()`.

MaanooAk commented on src/FreeAge/common/player.cpp:
```diff
@@ -67,65 +72,68 @@ void PlayerStats::BuildingFinished(BuildingType buildingType) {
-  // log();
+void PlayerStats::Change(const ObjectTypeStats& stats, int d) {
+  // Keep track of the population. The population values can only change in the start of the 
+  // game, so no total re-evaluations are needed.
```
I gave it an other try...

puzzlepaint replied:
Thanks, now I understand it.

MaanooAk commented on src/FreeAge/server/game.hpp:
```diff
@@ -152,9 +145,10 @@ class Game {
 
   /// The Gaia's stats
-  PlayerStats gaiaStats;
+  Player gaiPlayer;
```
Same old...

MaanooAk commented:
Went back to the PopulationCount and I added some comments and found the hidden `GetMoveSpeed`...

puzzlepaint commented:
Thanks for the update, looks good to me!
